gateway-service/src/main/java/org/zowe/apiml/gateway/conformance/AbstractSwaggerValidator.java

```diff
@@ -0,0 +1,114 @@
+ */
+
+package org.zowe.apiml.gateway.conformance;
+
+
+import org.springframework.http.HttpMethod;
+import org.zowe.apiml.product.gateway.GatewayConfigProperties;
+
+import java.util.*;
+
+public abstract class AbstractSwaggerValidator {
+
+    protected final Map<String, String> metadata;
+    protected final GatewayConfigProperties gatewayConfigProperties;
+    protected final String serviceId;
+
+    protected AbstractSwaggerValidator(Map<String, String> metadata, GatewayConfigProperties gatewayConfigProperties, String serviceId) {
+        this.metadata = metadata;
+        this.gatewayConfigProperties = gatewayConfigProperties;
+        this.serviceId = serviceId;
+    }
+
+    /**
+     * Returns list of issues with compliance with the OpenAPI specification
+     */
+    abstract List<String> getMessages();
+
+    /**
+     * Returns list of all documented endpoints
+     */
+    abstract Set<Endpoint> getAllEndpoints();
+
+    /**
+     * Returns list of issues with the endpoint URLs, mainly regarding the versioning part of conformance criteria
+     */
+    public List<String> getProblemsWithEndpointUrls() {
+        Set<Endpoint> endpoints = getAllEndpoints();
+
+        ArrayList<String> result = new ArrayList<>();
+
+        boolean isVersionedApi = !searchMetadata(metadata, "apiml", "apiInfo", "version").isEmpty();
+
+        for (Endpoint endpoint : endpoints) {
+            if (endpoint.getHttpMethods().isEmpty()) {
+                result.add("Endpoint at " + result + " does not have a specified Http method in Swagger");
+            }
+            if (!endpoint.getUrl().contains("/api/")) {
+                result.add("REST endpoint at " + result + " is not routed according to item 8 of the conformance criteria, missing /api/");
+                continue;
+            }
+            if (isVersionedApi) {
+                int start = endpoint.getUrl().indexOf("/api/") + 5; //this is where the version starts
+                String version;
+                try {
+                    version = endpoint.getUrl().substring(start);
+                    version = version.substring(0, version.indexOf("/"));
+                } catch (IndexOutOfBoundsException e) {
+                    version = "";
+                }
+
+                if (!(version.startsWith("v") && isNumber(version.substring(1)))) {
+                    result.add("REST endpoint at " + result + " is not versioned according to item 8 of the conformance criteria");
+                }
+            }
+        }
+        return result;
+    }
+
+    public Set<Endpoint> getGetMethodEndpoints() {
+        Set<Endpoint> result = new HashSet<>();
+
+        for (Endpoint endpoint : getAllEndpoints()) {
+            if (endpoint.getHttpMethods().contains(HttpMethod.GET)) {
+                result.add(endpoint);
+            }
+        }
+        return result;
+    }
+
+    public String searchMetadata(Map<String, String> metadata, String... partsOfKey) {
```

Comment:
Can we not send a list instead of arguments and use containsAll() from the ArrayList instead of defining a new method containsAll() here?

Reply:
We can always cast a list to array and pass it as an argument. 
The keys in the map can be of the form `apiml.apiInfo.0.swaggerUrl` or `apiml.apiInfo.api-v1.swaggerUrl` for example, I'd prefer using contains() on the whole key, instead of splicing it on the `.`